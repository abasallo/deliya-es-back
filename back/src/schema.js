import { gql } from 'apollo-server'

export default gql`
  schema {
    query: Query
    mutation: Mutation
  }

  type Query {
    users: [User]!
    userById(id: ID!): User
    userByUsername(username: String): User
  }

  type Mutation {
    addUser(user: UserInput!): User
  }

  type User {
    id: ID!
    username: String
    password: String
  }

  input UserInput {
    username: String
    password: String
  }
`
