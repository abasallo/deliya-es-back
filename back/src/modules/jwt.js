import jwt from 'jsonwebtoken'

const getUserFromToken = token => (token ? jwt.verify(token, process.env.JWT_SECRET).email : '')

const getTokenFromRequest = request => (request.headers.authorization || '').split(' ')[1]

export const getAuthenticatedUserFromRequest = request => getUserFromToken(getTokenFromRequest(request))
