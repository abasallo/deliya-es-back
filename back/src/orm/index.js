import Sequelize from 'sequelize'

const dbDialect = process.env.DB_DIALECT || 'sqlite'
const dbPath = process.env.DB_PATH || 'deliya.sqlite'

const sequelize = new Sequelize({
  dialect: dbDialect,
  storage: dbPath,
  pool: { max: 5, min: 0, idle: 10000 }
})

export const User = sequelize.define('user', {
  names: { type: Sequelize.STRING },
  surnames: { type: Sequelize.STRING },
  email: { type: Sequelize.STRING },
  password: { type: Sequelize.STRING },
  isEmailContactAllowed: { type: Sequelize.BOOLEAN }
})

// TODO - Refactor with async/await
export const initSequelize = () => {
  sequelize
    .authenticate()
    .then(() => {
      console.log('Connection has been established successfully.')
      User.sync({ force: true }).then(() =>
        User.create({
          names: 'Álvaro',
          surnames: 'Basallo Martínez',
          email: 'alvaro@basallo.es',
          password: '$2a$10$szbU0ZiERI8wFrbOaaTUnOqzkzKJAA4EJl6qRfGRZ8Moi07zipxTm',
          isEmailContactAllowed: true
        })
      )
    })
    .catch(err => console.error('Unable to connect to the database:', err))
}
