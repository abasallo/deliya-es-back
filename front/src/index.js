import React from 'react'
import ReactDOM from 'react-dom'

import { ApolloClient } from 'apollo-client'
import { createHttpLink } from 'apollo-link-http'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { ApolloProvider } from 'react-apollo'

import gql from 'graphql-tag'

import './index.css'

import * as serviceWorker from './serviceWorker'

import App from './App'

const client = new ApolloClient({ link: createHttpLink({ uri: 'http://localhost:4000/graphql' }), cache: new InMemoryCache() })

const query = gql`
  query {
    eventStates {
      id
      picture
      name
      description
      map {
        id
        title
        src
      }
      todoList {
        id
        primary
        secondary
      }
      productList {
        id
        primary
        secondary
      }
    }
  }
`

client.query({ query }).then(response => {
  ReactDOM.render(
    <ApolloProvider client={client}>
      <App eventState={response.data.eventStates[0]} />
    </ApolloProvider>,
    document.getElementById('root')
  )
})

// If you want your app to work offline and load faster, you can change unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister()
