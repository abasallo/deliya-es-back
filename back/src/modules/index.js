import jwt from 'jsonwebtoken'

export const getUserFromToken = token => (token ? jwt.verify(token, process.env.JWT_SECRET).email : '')

export const getTokenFromRequest = request => (request.headers.authorization || '').split(' ')[1]
