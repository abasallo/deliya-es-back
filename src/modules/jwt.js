import jwt from 'jsonwebtoken'

export const getUserFromToken = async token => {
  try {
    const decoded = await jwt.verify(token, process.env.JWT_SECRET)
    return decoded.email
  } catch (error) {
    return undefined
  }
}

export const getTokenFromRequest = request => {
  try {
    return request.headers.authorization.split(' ')[1]
  } catch (error) {
    return undefined
  }
}

export const getAuthenticatedUserFromRequest = request => getUserFromToken(getTokenFromRequest(request))

export const generateTokenFromEmailAndTTL = (email, ttl) =>
  jwt.sign({ email: email, date: Date.now() }, process.env.JWT_SECRET, { expiresIn: ttl || '10m' })

export const isTokenValid = token => jwt.verify(token, process.env.JWT_SECRET)
