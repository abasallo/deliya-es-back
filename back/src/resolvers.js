import eventStates from './data/eventStates'

import { pushObjectIntoArrayWithReturn } from './modules/utils'

export default {
  Query: {
    eventStates: () => eventStates,
    eventState: (parent, args) => eventStates.find(_ => _.id === args.id)
  },

  Mutation: {
    addBenchmarking: (parent, args) =>
      pushObjectIntoArrayWithReturn(
        {
          id: Date.now().toString(),
          picture: args.eventState.picture,
          name: args.eventState.name,
          description: args.eventState.description,
          mapId: args.eventState.mapId,
          todoList: args.eventState.todoList,
          productList: args.eventState.productList
        },
        eventStates
      )
  }
}
