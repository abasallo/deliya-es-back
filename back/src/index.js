import { ApolloServer } from 'apollo-server'

import typeDefs from './schema'

import resolvers from './resolvers'

import { initSequelize } from './orm'

import dotenv from 'dotenv'

import { getTokenFromRequest, getUserFromToken } from './modules'

dotenv.config()

initSequelize()

new ApolloServer({
  typeDefs,
  resolvers,
  context: ({ req }) => ({ user: getUserFromToken(getTokenFromRequest(req)) })
})
  .listen({ port: process.env.PORT || 4000 })
  .then(({ url }) => console.log(`Server ready at ${url}`))

// TODO - Too much information on Authentication errors is returned, is OK in DEV, but should be 401 in PROD
