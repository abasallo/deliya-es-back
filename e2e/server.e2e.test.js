import { model, server } from '../src/server-helper'

import { createApolloFetch } from 'apollo-fetch'

import jwt from 'jsonwebtoken'

let fetch
let apolloServer
beforeAll(async done => {
  fetch = await createApolloFetch({ uri: `http://${process.env.BACKEND_HOST}:${process.env.BACKEND_PORT}/graphql` })
  await model.then(() =>
    server.listen({ port: process.env.BACKEND_PORT }).then(_ => {
      apolloServer = _
      done()
    })
  )
})

afterAll(async () => await apolloServer.server.close())

test('Login with user and password', async () => {
  expect(
    await fetch({
      query: 'query($email: String, $password: String) { login(email: $email, password: $password) }',
      variables: { email: 'user@host.tld', password: 'password' }
    })
  ).toMatchSnapshot({ data: { login: expect.any(String) } })
})

test('Login with user and incorrect password', async () => {
  expect(
    await fetch({
      query: 'query($email: String, $password: String) { login(email: $email, password: $password) }',
      variables: { email: 'user@host.tld', password: 'wrongPassword' }
    })
  ).toMatchSnapshot()
})

test('Request password recovery', async () => {
  expect(
    await fetch({
      query: 'query($email: String) { requestPasswordRecoveryUrlOverEmail(email: $email) }',
      variables: { email: 'user@host.tld' }
    })
  ).toMatchSnapshot()
})

test('Request password recovery with nonexistent email', async () => {
  expect(
    await fetch({
      query: 'query($email: String) { requestPasswordRecoveryUrlOverEmail(email: $email) }',
      variables: { email: 'inexistentUser@host.tld' }
    })
  ).toMatchSnapshot()
})

test('Add user', async () => {
  expect(
    await fetch({
      query: `mutation($names: String, $surnames: String, $email: String!, $password: String!, $isEmailContactAllowed: Boolean!) {
                addUser(user: { names: $names, surnames: $surnames, email: $email, password: $password, isEmailContactAllowed: $isEmailContactAllowed }) {
                  id
                  names
                  surnames
                  email
                  password
                  isEmailContactAllowed
                }
      }`,
      variables: { names: 'names', surnames: 'surnames', email: 'email', password: 'password', isEmailContactAllowed: true }
    })
  ).toMatchSnapshot({ data: { addUser: { id: expect.any(String), password: expect.any(String) } } })
})

test('Change password with token', async () => {
  const token = await jwt.sign({ email: 'alvaro@basallo.es', date: Date.now() }, process.env.JWT_SECRET, { expiresIn: '10m' })
  expect(
    await fetch({
      query: 'mutation($password: String, $token: String) { changePasswordWithToken(password: $password, token: $token) }',
      variables: { password: 'newPassword', token: token }
    })
  ).toMatchSnapshot()
})

test('Change password with wrong, or expired, token', async () => {
  expect(
    await fetch({
      query: 'mutation($password: String, $token: String) { changePasswordWithToken(password: $password, token: $token) }',
      variables: { password: 'newPassword', token: 'wrong' }
    })
  ).toMatchSnapshot()
})
