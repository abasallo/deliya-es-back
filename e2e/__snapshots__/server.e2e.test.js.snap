// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activate User with token 1`] = `
Object {
  "data": Object {
    "activateUser": true,
  },
}
`;

exports[`Activate User with wrong, or expired, token 1`] = `
Object {
  "data": Object {
    "activateUser": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "message": "jwt malformed",
          "name": "JsonWebTokenError",
        },
      },
      "locations": Array [
        Object {
          "column": 28,
          "line": 1,
        },
      ],
      "message": "jwt malformed",
      "path": Array [
        "activateUser",
      ],
    },
  ],
}
`;

exports[`Add user 1`] = `
Object {
  "data": Object {
    "addUser": Object {
      "email": "email",
      "id": Any<String>,
      "isEmailContactAllowed": true,
      "names": "names",
      "password": Any<String>,
      "surnames": "surnames",
    },
  },
}
`;

exports[`Change password with token 1`] = `
Object {
  "data": Object {
    "changePasswordWithToken": true,
  },
}
`;

exports[`Change password with wrong, or expired, token 1`] = `
Object {
  "data": Object {
    "changePasswordWithToken": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "message": "jwt malformed",
          "name": "JsonWebTokenError",
        },
      },
      "locations": Array [
        Object {
          "column": 47,
          "line": 1,
        },
      ],
      "message": "jwt malformed",
      "path": Array [
        "changePasswordWithToken",
      ],
    },
  ],
}
`;

exports[`Check existing user existence 1`] = `
Object {
  "data": Object {
    "doesUserExists": true,
  },
}
`;

exports[`Check non existing user existence 1`] = `
Object {
  "data": Object {
    "doesUserExists": false,
  },
}
`;

exports[`Login with user and incorrect password 1`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 44,
          "line": 1,
        },
      ],
      "message": "Invalid password",
      "path": Array [
        "login",
      ],
    },
  ],
}
`;

exports[`Login with user and password 1`] = `
Object {
  "data": Object {
    "login": Any<String>,
  },
}
`;

exports[`Login with user and password, but deactivated 1`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "PERSISTED_QUERY_NOT_FOUND",
      },
      "locations": Array [
        Object {
          "column": 44,
          "line": 1,
        },
      ],
      "message": "PersistedQueryNotFound",
      "path": Array [
        "login",
      ],
    },
  ],
}
`;

exports[`Request password recovery 1`] = `
Object {
  "data": Object {
    "requestPasswordRecoveryUrlOverEmail": true,
  },
}
`;

exports[`Request password recovery with nonexistent email 1`] = `
Object {
  "data": Object {
    "requestPasswordRecoveryUrlOverEmail": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "PERSISTED_QUERY_NOT_FOUND",
      },
      "locations": Array [
        Object {
          "column": 25,
          "line": 1,
        },
      ],
      "message": "PersistedQueryNotFound",
      "path": Array [
        "requestPasswordRecoveryUrlOverEmail",
      ],
    },
  ],
}
`;

exports[`Request user activation 1`] = `
Object {
  "data": Object {
    "requestUserActivationUrlOverEmail": true,
  },
}
`;

exports[`Request user activation with nonexistent email 1`] = `
Object {
  "data": Object {
    "requestUserActivationUrlOverEmail": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "PERSISTED_QUERY_NOT_FOUND",
      },
      "locations": Array [
        Object {
          "column": 25,
          "line": 1,
        },
      ],
      "message": "PersistedQueryNotFound",
      "path": Array [
        "requestUserActivationUrlOverEmail",
      ],
    },
  ],
}
`;
